### PR TITLE
PR: Ask the frontend to save files before running them

### PR DESCRIPTION
**Pull request by impact27**
Give a chance to the frontend to save a file before running it.

This is the companion of spyder-ide/spyder#9966.

### CODE REVIEW

**Comment by ccordoba12** [comment]
@impact27, please remove all merges in this branch and leave a single one with master.

File: spyder_kernels/comms/commbase.py [hunk 1]
```diff
@@ -455,7 +455,7 @@ def _async_error(self, error_wrapper):
         """
         Handle an error that was raised on the other side asyncronously.
         """
-        error_wrapper.raise_error()
+        error_wrapper.print_error()
```
Comment (ccordoba12) [inline spyder_kernels/comms/commbase.py]: Why this change?

Comment (impact27) [reply spyder_kernels/comms/commbase.py]: I don't want to raise the error for async call, because if the call is async, the caller will not be expecting an exception. This is only a problem on the kernel side as the frontend reimplements the function.

Comment (impact27) [reply spyder_kernels/comms/commbase.py]: For example, if I make a syncronous call, I don't want to have an error for an unrelated async call being raised.

Comment (impact27) [reply spyder_kernels/comms/commbase.py]: Thinking about it, I have the following question:
The kernel makes a call that fails on the frontend side. Where should the error go?
 - If the error is syncronous, it should go back to the kernel side, because the calling function might want to handle the error.
 - If the call is asyncronous, the kernel side should be notified as well, so the callback for this call can be discarded.

But should we send back async error to the frontend to be displayed in the error window?

Comment (impact27) [reply spyder_kernels/comms/commbase.py]: I can do that easily. Should I open a new PR or attach this to this one?

Comment (ccordoba12) [reply spyder_kernels/comms/commbase.py]: > But should we send back async error to the frontend to be displayed in the error window?

Yes, I think that's important to do it.

> Should I open a new PR or attach this to this one?

Please do it here (this PR is already very simple).

File: spyder_kernels/customize/spydercustomize.py [hunk 2]
```diff
@@ -839,6 +839,11 @@ def runfile(filename, args=None, wdir=None, namespace=None, post_mortem=False):
     wdir: working directory
     post_mortem: boolean, whether to enter post-mortem mode on error
     """
+    try:
+        # Save the open files
+        _frontend_request().savefiles()
```
Comment (ccordoba12) [inline spyder_kernels/customize/spydercustomize.py]: I think this should contain `filename` as part of its arguments.

Comment (ccordoba12) [reply spyder_kernels/customize/spydercustomize.py]: I mean, because the editor needs to verify if the file is open there before trying to save it, right?

Comment (impact27) [reply spyder_kernels/customize/spydercustomize.py]: Well for now the frontend saves all the files when you run a script (eg. by pressing the run button). This makes sense because one file can call another. This just uses this already implemented mechanism. I can add the file name as an argument but it won't be used. 
For ref:
https://github.com/spyder-ide/spyder/blob/82cad82ef1fa454fa0703ce6db90609ab31784dd/spyder/plugins/editor/plugin.py#L2330

Comment (impact27) [reply spyder_kernels/customize/spydercustomize.py]: And you can run a file that is not open.
